lib/cloud/clients.go

```diff
@@ -231,143 +234,133 @@ type azureClients struct {
 	azureRunCommandClients azure.ClientMap[azure.RunCommandClient]
 }
 
-// GetAWSSession returns AWS session for the specified region.
-func (c *cloudClients) GetAWSSession(region string) (*awssession.Session, error) {
-	c.mtx.RLock()
-	if session, ok := c.awsSessions[region]; ok {
-		c.mtx.RUnlock()
-		return session, nil
-	}
-	c.mtx.RUnlock()
-	return c.initAWSSession(region)
+type SessionOptions struct {
+	Region           string
+	Session          *awssession.Session
+	RoleARN          string
+	ExternalID       string
+	CacheAssumedRole bool
 }
 
-// GetAWSSessionForRole returns AWS session for the specified role ARN.
-func (c *cloudClients) GetAWSSessionForRole(ctx context.Context, region, roleARN string) (*awssession.Session, error) {
-	defaultSession, err := c.GetAWSSession(region)
-	if err != nil {
-		return nil, trace.Wrap(err)
+func SessionOptionsFromAWSMeta(meta types.AWS) SessionOptions {
+	return SessionOptions{
+		Region:           meta.Region,
+		RoleARN:          meta.AssumeRoleARN,
+		ExternalID:       meta.ExternalID,
+		CacheAssumedRole: true,
 	}
+}
 
-	// Make a credentials with AssumeRoleProvider and test it out.
-	cred := stscreds.NewCredentials(defaultSession, roleARN)
-	if _, err := cred.GetWithContext(ctx); err != nil {
-		return nil, trace.Wrap(libcloudaws.ConvertRequestFailureError(err))
+func (c *cloudClients) GetAWSSession(ctx context.Context, options SessionOptions) (*awssession.Session, error) {
+	var err error
+	if options.Session == nil {
+		options.Session, err = c.getAWSSessionForRegion(options.Region)
```

Comment:
This is quite unintuitive that `GetAWSSession` returns the session put also modify the input parameter and set options.session = `newsession` and also returns the `newsession`.

Reply:
ok, the input parameter is passed by copy so it doesn't affect caller's SessionOption. It was just a convenience to reuse the options

Reply:
with refactor to use option funcs, this no longer modifies the input parameter either. Semantically it makes no difference, but maybe this is more intuitive? I declare a local variable instead and use/modify that.